src/java.base/share/classes/jdk/internal/foreign/StringSupport.java

```diff
@@ -93,15 +118,83 @@ private static void writeFast_int(MemorySegment segment, long offset, Charset ch
+    private static int native_strlen_byte(MemorySegment segment, long start) {
+        if (start > 0) {
+            segment = segment.asSlice(start);
+        }
+        long segmentSize = segment.byteSize();
+        final long len;
+        if (SIZE_T_IS_INT) {
+            if (segmentSize < MAX_TRIVIAL_SIZE) {
+                len = strnlen_int_trivial(segment, segmentSize);
+            } else if (segmentSize < Integer.MAX_VALUE * 2L) { // size_t is unsigned
+                len = strnlen_int(segment, segmentSize);
+            } else {
+                // There is no way to express the max size in the native method using an int so, revert
+                // to a Java method. It is possible to use a reduction of several STRNLEN invocations
+                // in a future optimization.
+                len = strlen_byte(segment);
+            }
+        } else {
+            len = segmentSize < MAX_TRIVIAL_SIZE
+                    ? strnlen_long_trivial(segment, segmentSize)
+                    : strnlen_long(segment, segmentSize);
+        }
+        if (len > ArraysSupport.SOFT_MAX_ARRAY_LENGTH) {
+            throw newIaeStringTooLarge();
+        }
+        return (int)len;
+    }
+
+    static long strnlen_int_trivial(MemorySegment segment, long size) {
+        try {
+            return Integer.toUnsignedLong((int)STRNLEN_TRIVIAL.invokeExact(segment, (int)size));
+        } catch (RuntimeException | Error e) {
+            throw e;
+        } catch (Throwable e) {
+            throw new IllegalArgumentException(e);
+        }
+    }
+
+    static long strnlen_int(MemorySegment segment, long size) {
+        try {
+            return Integer.toUnsignedLong((int)STRNLEN.invokeExact(segment, (int)size));
+        } catch (RuntimeException | Error e) {
+            throw e;
+        } catch (Throwable e) {
+            throw new IllegalArgumentException(e);
+        }
+    }
+
+    static long strnlen_long_trivial(MemorySegment segment, long size) {
+        try {
+            return (long)STRNLEN_TRIVIAL.invokeExact(segment, size);
+        } catch (RuntimeException | Error e) {
+            throw e;
+        } catch (Throwable e) {
+            throw new IllegalArgumentException(e);
+        }
+    }
+
+    static long strnlen_long(MemorySegment segment, long size) {
+        try {
+            return (long)STRNLEN.invokeExact(segment, size);
+        } catch (RuntimeException | Error e) {
+            throw e;
+        } catch (Throwable e) {
+            throw new IllegalArgumentException(e);
+        }
+    }
+
+    private static int strlen_byte(MemorySegment segment) {
         // iterate until overflow (String can only hold a byte[], whose length can be expressed as an int)
-        for (int offset = 0; offset >= 0; offset++) {
-            byte curr = segment.get(JAVA_BYTE, start + offset);
+        for (int offset = 0; offset >= 0; offset += 2) {
+            short curr = segment.get(JAVA_SHORT, offset);
```

Comment:
The old version intentionally used `JAVA_BYTE` here, as `strlen` and `strnlen` is for byte‑encoded strings (e.g.: **UTF‑8**):
```suggestion
        for (int offset = 0; offset >= 0; offset++) {
            short curr = segment.get(JAVA_BYTE, offset);
```

Reply:
We are planning on a new PR that is using another algorithm.